### ISSUE TITLE
Support to cuddle assignments for a whole block

### ISSUE THREAD

**Issue by bombsimon** [bombsimon/wsl#24]
Right now you have to use the variable, type or field in the first statement in a block if you want to use it like this:

```go
variable := "assigned"
[token] {
    // Must use 'variable' here
}
```

I think this is reasonable for most of the time but I kind of like to be able to use this pattern, especially if it's spawning a go routine like this:

```go
done := make(chan struct{})
go func() {
    fmt.Println("Let's spawn this go routine")
    <-somethingBlocking

    close(done)
}()

// Wait for done
<-done
```

This _might_ be useful in more scenarios like ranges, if-statements etcetera and therefore could possible be configured. Although as of now the only place I really felt that this was needed was a very few go routines.

Since this linter isn't really embrarcing any kind of **official** best practices like pythons black I think it's ok to have the linter heavily configurable like e.g. perltidy.

**Comment by hollowaykeanho** [comment]
Don't quite get you.. can you derive an example based on the former, using the latter's contents?

**Comment by bombsimon** [comment]
The idea is to be able to use the cuddled variable wherever you want in the block. Instead of just this:

```go
counter := 0
if somethingTrue {
    counter++
}
```

I also think it should be possible to configure it like this:
```go
counter := 0
if somethingTrue {
    checker := getAChecker()
    if !checker {
        return
    }

    counter++ // Cuddled variable used in block, but not as first statement
}
```

**Comment by hollowaykeanho** [comment]
>```
> counter := 0
> if somethingTrue {
>    counter++
> }
>```

+1. The rule between the assignment and definition cuddling (related) is very strict at the moment. 

> counter := 0
> if somethingTrue {
>    checker := getAChecker()
>    if !checker {
>        return
>    }
>
>    counter++ // Cuddled variable used in block, but not as first statement
>}
>```

I don't see any difference at the moment but this one is visually and cognitively acceptable.

**Comment by bombsimon** [comment]
I'm not really sure what the best way to attack this issue would be. Since this should be valid for every block I need to store all variables from all blocks in `parseBlockStatements` and return them to the caller of the function. However, right now I'm not parsing **every** line in a block but only if it's cuddled with something and not the first block.

If I'm going to store all assigned/used variables from all blocks I think I want to re-think the current layout first.

**Comment by mitar** [comment]
Yes, cuddling of variable assignments with a followup block should be possible if they are then used somewhere (anywhere) inside the block. To me this is the main obstacle with wsl and it makes code less readable by requiring a newline in there.

**Comment by treuherz** [comment]
If it isn't likely this improvement will be made soon, would you be open to PR making it possible to toggle the assignments-cuddling-loops lint off?

**Comment by bombsimon** [comment]
> If it isn't likely this improvement will be made soon, would you be open to PR making it possible to toggle the assignments-cuddling-loops lint off?

Yeah I don't see a problem adding an opt-in to disable this, we already added (perhaps with _too_ little consideration) other exclude rules.

If we do this however I think it should be consistent and disabel the check for
* [if](https://github.com/bombsimon/wsl/blob/master/doc/rules.md#if-statements-should-only-be-cuddled-with-assignments-used-in-the-if-statement-itself) ([code](https://github.com/bombsimon/wsl/blob/4740984d9e36d97979ec59e51959d85190e77190/wsl.go#L448))
* [for](https://github.com/bombsimon/wsl/blob/master/doc/rules.md#for-statements-should-only-be-cuddled-with-assignments-used-in-the-iteration) ([code](https://github.com/bombsimon/wsl/blob/4740984d9e36d97979ec59e51959d85190e77190/wsl.go#L684))
* [range](https://github.com/bombsimon/wsl/blob/master/doc/rules.md#ranges-should-only-be-cuddled-with-assignments-used-in-the-iteration) ([code](https://github.com/bombsimon/wsl/blob/4740984d9e36d97979ec59e51959d85190e77190/wsl.go#L602))
* [switch](https://github.com/bombsimon/wsl/blob/master/doc/rules.md#switch-statements-should-only-be-cuddled-with-variables-switched) ([code](https://github.com/bombsimon/wsl/blob/4740984d9e36d97979ec59e51959d85190e77190/wsl.go#L732))
* [type switch](https://github.com/bombsimon/wsl/blob/master/doc/rules.md#type-switch-statements-should-only-be-cuddled-with-variables-switched) ([code](https://github.com/bombsimon/wsl/blob/4740984d9e36d97979ec59e51959d85190e77190/wsl.go#L745))
* [defer](https://github.com/bombsimon/wsl/blob/master/doc/rules.md#defer-statements-should-only-be-cuddled-with-expressions-on-same-variable) ([code](https://github.com/bombsimon/wsl/blob/4740984d9e36d97979ec59e51959d85190e77190/wsl.go#L613), special `FuncLit` [code](https://github.com/bombsimon/wsl/blob/4740984d9e36d97979ec59e51959d85190e77190/wsl.go#L663-L675))

While looking at this I saw that all of these allow cuddling with variable that's used _first in block_ except `switch` for some reason.

**Comment by treuherz** [comment]
Is there a distinction here between variables used in the opening statement or within the block itself? Does wsl currently treat these differently?

**Comment by bombsimon** [comment]
> Is there a distinction here between variables used in the opening statement or within the block itself? Does wsl currently treat these differently?

Yes it is, that's what mentioned in the ticket. F.ex. these are allowed:

```go
# Used in if/for/range
x := "x"
if x != "" {
    // ...
}

# Used in first block statement
x := "x"
if y {
    x = z
}

# Used as argument in first statement
x := "x"
if y {
    pass(x)
}
```

When parsing block statements (statements that has a body) we start by grabbing the first statement at https://github.com/bombsimon/wsl/blob/4740984d9e36d97979ec59e51959d85190e77190/wsl.go#L253 and then using that to create a list of usages at https://github.com/bombsimon/wsl/blob/4740984d9e36d97979ec59e51959d85190e77190/wsl.go#L302-L308.

I think for the long term solution we probably want to return all variables used (not shadowed) in `parseBlockBody` and use those at all the places we currently use `calledOrAssignedFirstInBlock`. Thinking of this now, it might be possible to solve without too much work, I'll see if I can find some time to look into this.

**Comment by treuherz** [comment]
I'm sorry, I should have been clearer with my question. What I was trying to ask was whether you want whatever toggle(s) I add to cover all three of the cases you've described, or separately toggle case 1 (used in opening statement) and cases 2 and 3 (used in block).

**Comment by bombsimon** [comment]
Ah, I see! I think it would make sense to disable it completely, so meaning either you have to use it in the block or first in statement like now, or the lint is off and you can use it wherever.

**Comment by bombsimon** [comment]
Another one of those super old issues finally getting resolved, this time thanks to @jacobdrury! Thank you so much!! 🎉 😍 

If you want this now, you can install `wsl` from `master`: `go install github.com/bombsimon/wsl/v4/cmd/wsl@master`. 

However I'll release a new version and once dependabot picks it up add the new option to `golangci-lint`. Keeping this issue open until that's done.

**Comment by jacobdrury** [comment]
@bombsimon Looks like this was picked up by dependabot! https://github.com/golangci/golangci-lint/commit/bb4d6b74a4db6b0b60dd2b59ef6abec52c9ecd75

**Comment by bombsimon** [comment]
> @bombsimon Looks like this was picked up by dependabot! https://github.com/golangci/golangci-lint/commit/bb4d6b74a4db6b0b60dd2b59ef6abec52c9ecd75

Yep, perfect! It won't be released until v2.0 though. `golangci-lint` has frozen v1 so I'll keep this around until a v2 release.

**Comment by bombsimon** [comment]
`golangci-lint` [v2.0.0](https://github.com/golangci/golangci-lint/releases/tag/v2.0.0) is now released! Thanks again @jacobdrury
